backend/src/main/java/gov/cdc/usds/simplereport/service/DataHubUploaderService.java

```diff
@@ -0,0 +1,137 @@
+        }
+
+        // timestamp of highest matched entry, used for the next query. This is ridiculous, it cannot be right.
+        this._nextTimestamp = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS").format(events.get(0).getCreatedAt()).toString();
+        this._rowCount = events.size();
+
+        List<TestEventExport> eventsToExport = new ArrayList<>();
+        events.forEach(e -> eventsToExport.add(new TestEventExport(e)));
+
+        CsvMapper mapper = new CsvMapper();
+        mapper.enable(CsvGenerator.Feature.STRICT_CHECK_FOR_QUOTING)
+                .enable(CsvGenerator.Feature.ALWAYS_QUOTE_STRINGS)
+                .enable(CsvGenerator.Feature.ALWAYS_QUOTE_EMPTY_STRINGS);
+        // You would think `withNullValue` and `ALWAYS_QUOTE_EMPTY_STRINGS` would be enough, but you'd be wrong,
+        // we have to return `""` withNullValue to not put `,,,` in the csv
+        CsvSchema schema = mapper.schemaFor(TestEventExport.class).withHeader().withNullValue("\"\"");
+        this._fileContents = mapper.writer(schema).writeValueAsString(eventsToExport);
+    }
+
+    private void _uploadCSVDocument(final String apiKey) throws IOException, RestClientException {
+        ByteArrayResource contentsAsResource = new ByteArrayResource(this._fileContents.getBytes(StandardCharsets.UTF_8));
+
+        RestTemplate restTemplate = new RestTemplateBuilder(rt -> rt.getInterceptors().add((request, body, execution) -> {
+            HttpHeaders headers = request.getHeaders();
+            headers.setContentType(new MediaType("text", "csv"));
+            headers.add("x-functions-key", apiKey);
+            headers.add("client", "simple_report");
+            headers.add("x-api-version", API_VERSION);
+            return execution.execute(request, body);
+        })).build();
+
+        _uploadResultId = restTemplate.postForObject(DATAHUB_UPLOAD_URL, contentsAsResource, String.class);
+    }
+
+    public String creatTestCVSForDataHub(String lastEndCreateOn) {
+        try {
+            this._createTestEventCSV(lastEndCreateOn);
+            return this._fileContents;
+        } catch (IOException err) {
+            return err.toString();
+        } catch (NoResultException err) {
+            return "No matching results for the given startupdateby param";
+        }
+    }
+
+    // todo: Think through this transaction. In theory, this operation will have it's own table to
+    // track and log uploads.
+    // There is also the risk of the top action running multiple times concurrently.
+    // ultimately, it would be nice if each row had an ID that could be dedupped on the server.
+    @Transactional(readOnly = true)
+    public String uploadTestEventCVSToDataHub(final String apiKey, String lastEndCreateOn) {
+        try {
+            this._createTestEventCSV(lastEndCreateOn);
+            this._uploadCSVDocument(apiKey);
+
+            JSONObject resultJson = new JSONObject();
+            resultJson.put("result", "ok");
```

Comment:
This is the HTTP status—don't need to include it in the body.

Reply:
You are correct especially since it's hard coded to be "ok". Was thinking it could include more than just ok. I'll remove it for now.